test/Microsoft.Health.Fhir.Tests.E2E/Rest/Audit/AuditTests.cs

```diff
@@ -0,0 +1,375 @@
+        {
+            await ExecuteAndValidate(
+                client =>
+                {
+                    client.HttpClient.SetBearerToken("invalid");
+
+                    return Task.CompletedTask;
+                },
+                HttpStatusCode.Unauthorized);
+        }
+
+        [Fact]
+        public async Task GivenARequest_WhenValidAuthorizationTokenWithInvalidAudienceIsSupplied_ThenAuditLogEntriesShouldBeCreated()
+        {
+            await ExecuteAndValidate(
+                client => client.RunAsClientApplication(TestApplications.WrongAudienceClient),
+                HttpStatusCode.Unauthorized);
+        }
+
+        [Fact]
+        public async Task GivenAResource_WhenNotAuthorized_ThenAuditLogEntriesShouldBeCreated()
+        {
+            await ExecuteAndValidate(
+                client => client.RunAsClientApplication(TestApplications.NativeClient),
+                HttpStatusCode.Forbidden);
+        }
+
+        private async Task ExecuteAndValidate<T>(Func<Task<FhirResponse<T>>> action, string expectedAction, ResourceType expectedResourceType, Func<T, string> expectedPathGenerator, HttpStatusCode expectedStatusCode)
+            where T : Resource
+        {
+            if (!_fixture.IsUsingInProcTestServer)
+            {
+                // This test only works with the in-proc server with customized middleware pipeline.
+                return;
+            }
+
+            FhirResponse<T> response = null;
+
+            response = await action();
+
+            string correlationId = response.Headers.GetValues(RequestIdHeaderName).FirstOrDefault();
+
+            Assert.NotNull(correlationId);
+
+            var expectedUri = new Uri($"http://localhost/{expectedPathGenerator(response.Resource)}");
+
+            Assert.Collection(
+                _auditLogger.GetAuditEntriesByCorrelationId(correlationId),
+                ae => ValidateExecutingAuditEntry(ae, expectedAction, expectedUri, correlationId),
+                ae => ValidateExecutedAuditEntry(ae, expectedAction, expectedResourceType, expectedUri, expectedStatusCode, correlationId));
+        }
+
+        private async Task ExecuteAndValidate(Func<FhirClient, Task> clientSetup, HttpStatusCode expectedStatusCode)
+        {
+            if (!_fixture.IsUsingInProcTestServer || !_fixture.FhirClient.SecuritySettings.SecurityEnabled)
+            {
+                // This test only works with the in-proc server with customized middleware pipeline and when security is enabled.
+                return;
+            }
+
+            const string url = "Patient/123";
+
+            // Create a new client with no token supplied.
+            var client = new FhirClient(_fixture.CreateHttpClient(), ResourceFormat.Json);
+
+            await clientSetup(client);
+
+            FhirResponse<OperationOutcome> response = null;
+
+            try
+            {
+                await client.ReadAsync<Patient>(url);
```

Comment:
Should we Assert.Fail() after this, in case the call actually succeeds?  The test will still fail later with a null ref exception, but an explicit failure here might be clearer.

Reply:
Updated the code to make sure it throws exception.